### ISSUE TITLE
Node upgrades: provision new nodes

### ISSUE THREAD

**Issue by mbforbes** [GoogleCloudPlatform/kubernetes#6088]
See issue #6079 as the roll-up for node upgrades.

**tl;dr:**
Update MIG template, then call a MIG rollingupdate. 

**More details:**
When upgrading nodes, an in-place upgrade must worry about leftover state on the boot disk, VM configuration like IP tables, and the idempotency of salt. Getting a fresh VM instead would remove these concerns and provide a clean state for a new node.

Furthermore, a versioned node spec (#4855) includes the tuple of (kubelet, kube-proxy, docker, kernel/OS). In the general case, changing docker or the kernel/OS means we have to stop the running pods. Once know we have to stop the pods, the simplest upgrade mechanism is to totally down the node and provision a new one.

Provisioning a new node is also a great backup to have for node repairs. To be clear, though, this issue is entirely separate from node repairs; once we have the capability to grow the cluster by adding new nodes, this mechanism, and other repair mechanisms, will build on top of that capability.

This was blocked by:

- #6087 nodes joining a running cluster
- #7912 (template-only metadata)

**References:**
- The most relevant discussion regarding the worries of complications with state while doing in-place upgrades is #3333.

**Comment by alex-mohr** [comment]
FWIW, I think we need this as a fallback to repair completely borked nodes?  That is, it may not be necessary for upgrades per se, but could end up being independently useful.

**Comment by mbforbes** [comment]
Thanks for the comment, @alex-mohr. I updated the description to be clear about what this issue is addressing. Having the ability to add new nodes to a cluster will both enable this upgrade path as well as give us a great "hammer" to use for repairs.

**Comment by mbforbes** [comment]
Working on this now.
